### PR TITLE
drop support for unmaintained versions of Node.js; upgrade stuff

### PR DESCRIPTION
**Pull request by boneskull**
Addresses #3016

- Drops support for Node.js v0.10, v0.12, v5.x and iojs
- Address potential [arbitrary code injection vulnerability](https://snyk.io/vuln/npm:growl:20160721) in `growl`
- Address potential [regex DoS vulnerability](https://snyk.io/vuln/npm:debug:20170905) in `debug`
- Upgrade `eslint` and plugins and sweep up code
- Upgrade `should`, `growl`, and `debug`
- Update `engines` field in `package.json`

I'm going to see what else I can squeeze out of this.

### CODE REVIEW

**Comment by boneskull** [comment]
This will be (squashed) and merged when the build's green.

File: lib/runner.js [hunk 1]
```diff
@@ -229,7 +229,7 @@ Runner.prototype.fail = function (test, err) {
   ++this.failures;
   test.state = 'failed';
 
-  if (!(err instanceof Error || err && typeof err.message === 'string')) {
+  if (!((err instanceof Error || err) && typeof err.message === 'string')) {
```
Comment (ScottFreeCode) [inline lib/runner.js]: Are we sure this shouldn't be `if (!(err instanceof Error || (err && typeof err.message === 'string'))) {`? Reading the original, and not remembering exactly how JS evaluates differing chained boolean operators without parentheses, it *looks* like the original included `err` as a condition to guard against accessing `.message` of `undefined`.

File: lib/utils.js [hunk 2]
```diff
@@ -540,8 +540,8 @@ function jsonStringify (object, spaces, depth) {
     --length;
     str += '\n ' + repeat(' ', space) +
       (isArray(object) ? '' : '"' + i + '": ') + // key
-      _stringify(object[i]) +                    // value
-      (length ? ',' : '');                       // comma
+      _stringify(object[i]) + // value
+      (length ? ',' : ''); // comma
```
Comment (ScottFreeCode) [inline lib/utils.js]: [unimportant] I long for the day that linters can recognise when whitespace has been properly used to make things, any things, line up across multiple lines. [/unimportant]

File: lib/runner.js (hunk 1 above)
Comment (boneskull) [reply lib/runner.js]: I think you're right.

**Pull request merged**